src/neo/SmartContract/Native/NameService.cs

```diff
@@ -0,0 +1,247 @@
+#pragma warning disable IDE0051
+
+using Neo.Cryptography;
+using Neo.IO;
+using Neo.IO.Json;
+using Neo.Ledger;
+using Neo.Persistence;
+using Neo.VM;
+using Neo.VM.Types;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Numerics;
+using System.Text.RegularExpressions;
+
+namespace Neo.SmartContract.Native
+{
+    public sealed class NameService : NonfungibleToken<NameService.NameState>
+    {
+        public override int Id => -6;
+        public override string Symbol => "NNS";
+
+        private const byte Prefix_Roots = 10;
+        private const byte Prefix_DomainPrice = 22;
+        private const byte Prefix_Expiration = 20;
+        private const byte Prefix_Record = 12;
+
+        private const uint OneYear = 365 * 24 * 3600;
+        private static readonly Regex rootRegex = new Regex("^[a-z][a-z0-9]{0,15}$");
+        private static readonly Regex nameRegex = new Regex("^(?=.{3,255}$)([a-z0-9]{1,62}\\.)+[a-z][a-z0-9]{0,15}$");
+
+        internal NameService()
+        {
+        }
+
+        internal override void Initialize(ApplicationEngine engine)
+        {
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_DomainPrice), new StorageItem(10_00000000));
+        }
+
+        internal override void OnPersist(ApplicationEngine engine)
+        {
+            uint now = (uint)(engine.Snapshot.PersistingBlock.Timestamp / 1000) + 1;
+            byte[] start = CreateStorageKey(Prefix_Expiration).AddBigEndian(0).ToArray();
+            byte[] end = CreateStorageKey(Prefix_Expiration).AddBigEndian(now).ToArray();
+            foreach (var (key, _) in engine.Snapshot.Storages.FindRange(start, end))
```

Comment:
Maybe it is possible to just delete those expired within the last couple of seconds.
The start can be the last end.

Reply:
There is no difference when doing db search.